src/rust/catnap/socket.rs

```diff
@@ -62,107 +82,297 @@ pub struct Socket {
 
 impl Socket {
     /// Creates a new socket that is not bound to an address.
-    pub fn new() -> Self {
-        Self {
-            state: SocketState::NotBound,
-            local: None,
-            remote: None,
+    pub fn new(domain: libc::c_int, typ: libc::c_int) -> Result<Self, Fail> {
+        // Parse communication domain.
+        if domain != libc::AF_INET {
+            return Err(Fail::new(libc::ENOTSUP, "communication domain not supported"));
+        }
+
+        // Parse socket type and protocol.
+        if (typ != libc::SOCK_STREAM) && (typ != libc::SOCK_DGRAM) {
+            return Err(Fail::new(libc::ENOTSUP, "socket type not supported"));
         }
```

Comment:
We are not consistent on parameter checking logic. Some are now performed in `Socket`, others in `CatnapLibOS` and, yet others in both.

I have the following suggestion:
- Keep the parameter checking logic in `CatnapLibos` level.
- Add debug asserts on `Socket` level.

Reply:
Perfect, thanks!

Reply:
btw, I would like to keep everything that depends on libc out of CatnapLibOS. Will that still be possible?